blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py

```diff
@@ -0,0 +1,458 @@
+        unique_dates = df['image_upload_date'].unique()
+        logger.info(f'Unique dates: {unique_dates}')
+
+        for upload_date in unique_dates:
+            # Filter for the current date
+            df_date = df[df['image_upload_date'] == upload_date]
+            total_images = len(df_date)
+            logger.info(f'Total images on date {upload_date}: {total_images}')
+            
+            # Calculate the total number of images to sample
+            total_images_to_sample = int(total_images * percentage_ratio)
+            
+            # Ensure at least one image is sampled if total_images_to_sample > 0
+            total_images_to_sample = max(total_images_to_sample, 1) if total_images > 0 else 0
+            
+            logger.info(f'Total images to sample on date {upload_date}: {total_images_to_sample}')
+            
+            # Check if 'bin_label' column exists
+            if 'bin_label' not in df_date.columns:
+                logger.error("bin_label column not found in DataFrame for date: " + str(upload_date))
+                continue  # Skip to next date or handle error as needed
+
+            # Get the unique bin labels
+            bin_labels = df_date['bin_label'].unique()
+            logger.debug(f'Bin labels for date {upload_date}: {bin_labels}')
+
+            # Calculate the number of images to sample per bin
+            images_per_bin = total_images_to_sample // len(bin_labels)
+            remainder = total_images_to_sample % len(bin_labels)
+            logger.info(f'Images per bin: {images_per_bin}, Remainder: {remainder}')
+            
+            for bin_label in bin_labels:
+                df_bin = df_date[df_date['bin_label'] == bin_label]
+                
+                # Adjust the number of images to sample from this bin
+                num_to_sample = min(images_per_bin + (1 if remainder > 0 else 0), len(df_bin))
+                remainder -= 1 if remainder > 0 else 0
+                
+                # Sample images
+                sampled_from_bin = df_bin.sample(n=num_to_sample)
+                sampled_images_df = sampled_images_df.append(sampled_from_bin)
+                
+            # Print statement for debugging
+            logger.debug(f'Sampled images for date {upload_date}: {sampled_images_df.head()}')
+
+        # Return the full DataFrame after processing all dates
+        return sampled_images_df
+    
+    @staticmethod
+    def determine_bin_size(detection_range: int) -> int:
+        """
+        Determines the bin size for categorization based on the detection range.
+
+        Parameters
+        ----------
+        detection_range : int
+            The range of detection counts across all images.
+
+        Returns
+        -------
+        int
+            The number of bins to be used for categorization.
+        """
+        
+        if detection_range <= 10:
+            return 3
+        elif 10 < detection_range <= 50:
```

Comment:
else is irrelevant since there is a return, not an error but just a style thing :P
Also you could do:
```
if detection_range > 50:
   return 10
if detection_range > 10:
   return 5
return 3
```